### PR TITLE
Put global box length and periodicity into class, pass as argument

### LINKED ISSUE #2675: Output parameter version of `get_mi_vector` should be removed
to discourage bad code where the result is not `const`.

### PR DESCRIPTION
**Pull request by fweik**
Fixes #2675.

Description of changes:
 - Introduce type for global box geometry
- Pass box geometry as argument to some functions
- de-feature `PARTIAL_PERIODIC`

### CODE REVIEW

**Comment by fweik** [comment]
@RudolfWeeber @jngrad any input on why the virtual sites and bond angle stuff is not working anymore is welcome. It's not obvious to me.

File: src/core/electrostatics_magnetostatics/mmm2d.cpp [hunk 1]
```diff
@@ -295,6 +296,19 @@ void MMM2D_setup_constants() {
   }
 }
 
+static void layered_get_mi_vector(double res[3], double const a[3],
```
Comment (KaiSzuttor) [inline src/core/electrostatics_magnetostatics/mmm2d.cpp:299]: why did you move the function to mm2d?

Comment (fweik) [reply src/core/electrostatics_magnetostatics/mmm2d.cpp:299]: Because it's only used there

**Comment by fweik** [comment]
Sorry forgot something.

**Comment by jngrad** [comment]
@fweik I don't see what you mean by virtual sites and bonds, on my machine I can't even get the PR to compile: the Cython code is unable to cast `Vector3d` to `const Vector3d` nor to choose which overloaded `make_array_locked()` to import from `utils.pxd`.

**Comment by fweik** [comment]
Some of those are bugs in the cython compiler that are fixed in newer versions. Works here with cython 0.28.2 (from pip).

**Comment by fweik** [comment]
The doxygen warning seems to be a false positive to me, the cited line does not contain a `\ref` command.

**Comment by jngrad** [comment]
Let me have a look

**Comment by jngrad** [comment]
ah the prototype `static void dp3m_init_a_ai_cao_cut();` is declared twice at lines 98 and 131. The offending ref is in the first prototype. But both Doxygen comments don't match at all, one of them must be wrong.

**Comment by fweik** [comment]
Ah ok, I somehow failed to see how the error message is related to that...

**Comment by jngrad** [comment]
You still need to remove the `@ref` at line 96.

File: src/core/BoxGeometry.hpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+   *
+   * @param coord The coordinate to set the periodicity for.
+   * @param val True if this direction should be periodic.
+   */
+  void set_periodic(unsigned coord, bool val) { m_periodic.set(coord, val); }
+
+  /**
+   * @brief Check periodicity in direction.
+   *
+   * @param coord Direction to check
```
Comment (pkreissl) [inline src/core/BoxGeometry.hpp:27]: Typo "if".

File: src/core/BoxGeometry.hpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+   * @return true iff periodic in direction.
```
Comment (pkreissl) [inline src/core/BoxGeometry.hpp:28]: Comment refers to this line.

Comment (fweik) [reply src/core/BoxGeometry.hpp:28]: https://en.wikipedia.org/wiki/If_and_only_if :-)

Comment (pkreissl) [reply src/core/BoxGeometry.hpp:28]: Iff this was indeed intentional, leave as is ;-)

**Comment by fweik** [comment]
@KaiSzuttor @RudolfWeeber

**Review by jngrad — APPROVED** [review]

**Review by RudolfWeeber — APPROVED** [review]
